crates/orderbook/src/orderbook.rs

```diff
@@ -243,13 +274,68 @@ fn set_available_balances(orders: &mut [Order], cache: &SolvableOrdersCache) {
     }
 }
 
+async fn get_orders_with_native_prices(
+    mut orders: Vec<Order>,
+    native_price_estimator: &dyn NativePriceEstimating,
+) -> (Vec<Order>, BTreeMap<H160, U256>) {
+    let traded_tokens = orders
+        .iter()
+        .flat_map(|order| {
+            [
+                order.order_creation.sell_token,
+                order.order_creation.buy_token,
+            ]
+        })
+        .collect::<HashSet<_>>()
+        .into_iter()
+        .collect::<Vec<_>>();
+    let prices = native_price_estimator
+        .estimate_native_prices(&traded_tokens)
+        .await
+        .into_iter()
+        .zip(traded_tokens)
+        .filter_map(|(price, token)| match price {
+            Ok(price) => Some((token, to_normalized_price(price)?)),
+            Err(err) => {
+                tracing::warn!(?token, ?err, "error estimating native token price");
+                None
+            }
+        })
+        .collect::<BTreeMap<_, _>>();
+
+    orders.retain(|order| {
+        prices.contains_key(&order.order_creation.sell_token)
+            && prices.contains_key(&order.order_creation.buy_token)
+    });
+
+    (orders, prices)
+}
+
+fn to_normalized_price(price: f64) -> Option<U256> {
+    let uint_max = 2.0_f64.powi(256);
+
+    // NOTE: The `NativePriceEstimating` component returns prices denominated
+    // in the token that it is estimating and not in ETH. This means that we
+    // need to invert the price in order for it to be correct.
+    let price_in_eth = 1e18 / price;
+    if price_in_eth < 1. || price_in_eth >= uint_max {
+        return None;
+    }
```

Comment:
One more nit on this. Testing on my local machine 1e18 / 0.0 is INF. But I don't know if that is guaranteed. If the result was NAN then the if condition would be false and we would continue with U256::from_f64_lossy(NAN). So maybe should check for 0 price explicitly.

Reply:
According to the IEEE-754 standard page on Wikipedia:

> Division by zero: an operation on finite operands gives an exact infinite result, e.g., 1/0 or log(0). By default, returns ±infinity.

Note that `0. / 0. is NaN` though!

Reply:
Made this stricter. I additionally added a `price_in_eth.is_normal()` condition so that the price can't be `NaN`, zero, subnormal, or infinite.